client/src/js/directives/bhFindPatient.js

```diff
@@ -0,0 +1,263 @@
+angular.module('bhima.directives')
+.component('bhFindPatient', {
+  controller: FindPatientComponent,
+  templateUrl : 'partials/templates/findpatient.tmpl.html',
+  bindings: {
+    onSearchComplete: '&',  // bind callback
+    type:             '@',  // bind string
+    required:         '<',  // bind the required 
+  }
+});
+
+FindPatientComponent.$inject = ['Patients', 'appcache'];
+
+/**
+ * The Find Patient Component
+ *
+ * This component allows a user to serach for a patient by either the
+ * patient identifier (Project Abbreviation + Reference) or by typeahead on patient
+ * name.
+ *
+ * The typeahead loads data as your type into the input box, pinging th URL
+ * /patient/search/?name={string} The HTTP endpoints sends back 20 results
+ * which are presented to the user.
+ *
+ * SUPPORTED ATTRIBUTES:
+ *   - type : which take one of these values (inline or panel) (default: inline)
+ *   - on-search-complete : the callback function which get the returned patient
+ */
+function FindPatientComponent(Patients, AppCache) {
+  var vm = this;
+
+  /** cache to remember which the search type of the component */
+  var cache = new AppCache('FindPatientComponent');
+
+  /** @const the max number of records to fetch from the server */
+  var LIMIT = 20;
+
+  /** supported searches: by name or by id */
+  vm.options = [{
```

Comment:
I know this isn't your code (so I'm not sure where to post this) but this options configuration is actually quite difficult to follow - it seems to me that in a lot of places throughout this code we are relying on someone reading```selected.key === 0``` as 'find by ID state' and ```selected.key===1``` as 'find by name state'. This leads to a lot of constants being used throughout both the controller and the template. 

Potentially as a configuration object is being defined anyway would it be possible to do something like 
```js 
options = { 
  findById : { ... } // (label, placeholder etc.)
  findByName : { ... } // (label, placeholder etc.)
};
```

This might then read
```js
if (vm.selected === options.findById)
if (vm.selected === options.findByName)
```